File: agents/scala/fw-agent/src/main/scala/afl/agent/AgentPollerConfig.scala
package fw.agent

import java.net.InetAddress
import scala.io.Source
import scala.util.{Try, Using}

/** Configuration for the AFL agent poller.
  *
  * @param serviceName  Service identifier for server registration
  * @param serverGroup  Logical group name
  * @param serverName   Hostname (defaults to local hostname)
  * @param taskList     Task list name for routing
  * @param pollIntervalMs  Polling interval in milliseconds
  * @param maxConcurrent   Maximum concurrent event handlers
  * @param heartbeatIntervalMs  Heartbeat interval in milliseconds
  * @param mongoUrl    MongoDB connection string (required)
  * @param database    MongoDB database name
  */
case class AgentPollerConfig(
    serviceName: String = "fw-agent",
    serverGroup: String = "default",
    serverName: String = AgentPollerConfig.defaultHostname,
    taskList: String = "default",
    pollIntervalMs: Long = 2000,
    maxConcurrent: Int = 5,
    heartbeatIntervalMs: Long = 10000,
    mongoUrl: String,
    database: String = "afl"
)

object AgentPollerConfig:
  private[agent] def defaultHostname: String =
    Try(InetAddress.getLocalHost.getHostName).getOrElse("unknown")

  /** Load config from an afl.config.json file.
    *
    * Reads the mongodb.url, mongodb.database, and runner section fields.
    * Applies AFL_ENV overlay if set. Falls back to environment variables
    * and then built-in defaults.
    */
  def fromConfig(path: String): AgentPollerConfig =
    val content = Using(Source.fromFile(path))(_.mkString).getOrElse(
      throw new IllegalArgumentException(s"Cannot read config file: $path")
    )
    val base = fromJsonString(content)

    // AFL_ENV overlay
    sys.env.get("AFL_ENV").filter(_.nonEmpty) match
      case Some(envName) =>
        val dir = new java.io.File(path).getParent
        val overlayPath = s"$dir/afl.config.$envName.json"
        if new java.io.File(overlayPath).isFile then
          val overlayContent = Using(Source.fromFile(overlayPath))(_.mkString).getOrElse("")
          if overlayContent.nonEmpty then
            val overlay = fromJsonString(overlayContent)
            // Overlay values win over base when they differ from defaults
            base.copy(
              mongoUrl = if overlay.mongoUrl != "mongodb://localhost:27017" then overlay.mongoUrl else base.mongoUrl,
              database = if overlay.database != "afl" then overlay.database else base.database,
              pollIntervalMs = if overlay.pollIntervalMs != 2000L then overlay.pollIntervalMs else base.pollIntervalMs,
              maxConcurrent = if overlay.maxConcurrent != 5 then overlay.maxConcurrent else base.maxConcurrent,
              heartbeatIntervalMs = if overlay.heartbeatIntervalMs != 10000L then overlay.heartbeatIntervalMs else base.heartbeatIntervalMs
            )
          else base
        else base
      case None => base

  /** Parse config from a JSON string (afl.config.json content). */
  def fromJsonString(json: String): AgentPollerConfig =
    // Minimal JSON parsing — extract mongodb fields without a JSON library dependency.
    val url = extractField(json, "url")
      .orElse(sys.env.get("AFL_MONGODB_URL"))
      .getOrElse("mongodb://localhost:27017")
    val database = extractField(json, "database")
      .orElse(sys.env.get("AFL_MONGODB_DATABASE"))
      .getOrElse("afl")

    // Runner section fields
    val pollMs = extractIntField(json, "pollIntervalMs")
      .orElse(sys.env.get("AFL_POLL_INTERVAL_MS").flatMap(s => scala.util.Try(s.toLong).toOption))
      .getOrElse(2000L)
    val maxConc = extractIntField(json, "maxConcurrent")
      .orElse(sys.env.get("AFL_MAX_CONCURRENT").flatMap(s => scala.util.Try(s.toInt).toOption))
      .map(_.toInt)
      .getOrElse(5)
    val hbMs = extractIntField(json, "heartbeatIntervalMs")
      .orElse(sys.env.get("AFL_HEARTBEAT_INTERVAL_MS").flatMap(s => scala.util.Try(s.toLong).toOption))
      .getOrElse(10000L)

    AgentPollerConfig(
      mongoUrl = url,
      database = database,
      pollIntervalMs = pollMs,
      maxConcurrent = maxConc,
      heartbeatIntervalMs = hbMs
    )

  /** Resolve config using the standard search order from the protocol spec. */
  def resolve(explicitPath: Option[String] = None): AgentPollerConfig =
    val path = explicitPath
      .orElse(sys.env.get("AFL_CONFIG"))
      .orElse(findConfigFile())
    path match
      case Some(p) => fromConfig(p)
      case None    => fromEnvironment()

  /** Build config purely from environment variables. */
  def fromEnvironment(): AgentPollerConfig =
    AgentPollerConfig(
      mongoUrl = sys.env.getOrElse(
        "AFL_MONGODB_URL",
        "mongodb://localhost:27017"
      ),
      database = sys.env.getOrElse("AFL_MONGODB_DATABASE", "afl")
    )

  private def findConfigFile(): Option[String] =
    val candidates = Seq(
      "afl.config.json",
      sys.props.get("user.home").map(_ + "/.afl/afl.config.json").getOrElse(""),
      "/etc/afl/afl.config.json"
    ).filter(_.nonEmpty)
    candidates.find(p => new java.io.File(p).isFile)

  private[agent] def extractField(json: String, field: String): Option[String] =
    val pattern = s""""$field"\\s*:\\s*"([^"]+)"""".r
    pattern.findFirstMatchIn(json).map(_.group(1))

  private[agent] def extractIntField(json: String, field: String): Option[Long] =
    val pattern = s""""$field"\\s*:\\s*(\\d+)""".r
    pattern.findFirstMatchIn(json).flatMap(m => scala.util.Try(m.group(1).toLong).toOption)

File: agents/java/fw-agent/src/main/java/afl/agent/AgentPollerConfig.java
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fw.agent;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.net.InetAddress;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;

/**
 * Configuration for AgentPoller.
 */
public record AgentPollerConfig(
        String serviceName,
        String serverGroup,
        String serverName,
        String taskList,
        long pollIntervalMs,
        int maxConcurrent,
        long heartbeatIntervalMs,
        String mongoUrl,
        String database
) {

    /**
     * Returns a configuration with default values.
     */
    public static AgentPollerConfig defaults() {
        String hostname;
        try {
            hostname = InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            hostname = "unknown";
        }

        return new AgentPollerConfig(
                "fw-agent",
                "default",
                hostname,
                "default",
                2000,
                5,
                10000,
                "mongodb://localhost:27017",
                "afl"
        );
    }

    /**
     * Loads configuration from a JSON file.
     */
    public static AgentPollerConfig fromFile(String path) {
        AgentPollerConfig cfg = defaults();

        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(new File(path));

            JsonNode mongodb = root.get("mongodb");
            if (mongodb != null) {
                String url = Optional.ofNullable(mongodb.get("url"))
                        .map(JsonNode::asText)
                        .filter(s -> !s.isEmpty())
                        .orElse(cfg.mongoUrl());
                String database = Optional.ofNullable(mongodb.get("database"))
                        .map(JsonNode::asText)
                        .filter(s -> !s.isEmpty())
                        .orElse(cfg.database());

                cfg = new AgentPollerConfig(
                        cfg.serviceName(),
                        cfg.serverGroup(),
                        cfg.serverName(),
                        cfg.taskList(),
                        cfg.pollIntervalMs(),
                        cfg.maxConcurrent(),
                        cfg.heartbeatIntervalMs(),
                        url,
                        database
                );
            }

            // Runner section
            JsonNode runner = root.get("runner");
            if (runner != null) {
                long poll = Optional.ofNullable(runner.get("pollIntervalMs"))
                        .map(JsonNode::asLong)
                        .orElse(cfg.pollIntervalMs());
                int maxConc = Optional.ofNullable(runner.get("maxConcurrent"))
                        .map(JsonNode::asInt)
                        .orElse(cfg.maxConcurrent());
                long heartbeat = Optional.ofNullable(runner.get("heartbeatIntervalMs"))
                        .map(JsonNode::asLong)
                        .orElse(cfg.heartbeatIntervalMs());

                cfg = new AgentPollerConfig(
                        cfg.serviceName(),
                        cfg.serverGroup(),
                        cfg.serverName(),
                        cfg.taskList(),
                        poll,
                        maxConc,
                        heartbeat,
                        cfg.mongoUrl(),
                        cfg.database()
                );
            }

            // AFL_ENV overlay
            String envName = System.getenv("AFL_ENV");
            if (envName != null && !envName.isEmpty()) {
                File overlayFile = new File(new File(path).getParent(), "afl.config." + envName + ".json");
                if (overlayFile.exists()) {
                    try {
                        JsonNode overlay = mapper.readTree(overlayFile);
                        JsonNode oMongo = overlay.get("mongodb");
                        if (oMongo != null) {
                            String oUrl = Optional.ofNullable(oMongo.get("url"))
                                    .map(JsonNode::asText).filter(s -> !s.isEmpty()).orElse(cfg.mongoUrl());
                            String oDb = Optional.ofNullable(oMongo.get("database"))
                                    .map(JsonNode::asText).filter(s -> !s.isEmpty()).orElse(cfg.database());
                            cfg = new AgentPollerConfig(cfg.serviceName(), cfg.serverGroup(), cfg.serverName(),
                                    cfg.taskList(), cfg.pollIntervalMs(), cfg.maxConcurrent(),
                                    cfg.heartbeatIntervalMs(), oUrl, oDb);
                        }
                        JsonNode oRunner = overlay.get("runner");
                        if (oRunner != null) {
                            long oPoll = Optional.ofNullable(oRunner.get("pollIntervalMs"))
                                    .map(JsonNode::asLong).orElse(cfg.pollIntervalMs());
                            int oMaxConc = Optional.ofNullable(oRunner.get("maxConcurrent"))
                                    .map(JsonNode::asInt).orElse(cfg.maxConcurrent());
                            long oHb = Optional.ofNullable(oRunner.get("heartbeatIntervalMs"))
                                    .map(JsonNode::asLong).orElse(cfg.heartbeatIntervalMs());
                            cfg = new AgentPollerConfig(cfg.serviceName(), cfg.serverGroup(), cfg.serverName(),
                                    cfg.taskList(), oPoll, oMaxConc, oHb, cfg.mongoUrl(), cfg.database());
                        }
                    } catch (Exception ignored) {
                        // Overlay parse error — use base config
                    }
                }
            }
        } catch (Exception e) {
            // Ignore file read/parse errors, use defaults
        }

        return applyEnvOverrides(cfg);
    }

    /**
     * Resolves configuration using the standard search order:
     * 1. Explicit path argument
     * 2. AFL_CONFIG environment variable
     * 3. afl.config.json in current directory
     * 4. ~/.afl/afl.config.json
     * 5. /etc/afl/afl.config.json
     * 6. Environment variables
     * 7. Built-in defaults
     */
    public static AgentPollerConfig resolve(String explicitPath) {
        if (explicitPath != null && !explicitPath.isEmpty()) {
            if (Files.exists(Path.of(explicitPath))) {
                return fromFile(explicitPath);
            }
        }

        String envPath = System.getenv("AFL_CONFIG");
        if (envPath != null && !envPath.isEmpty() && Files.exists(Path.of(envPath))) {
            return fromFile(envPath);
        }

        String[] searchPaths = {
                "afl.config.json",
                System.getProperty("user.home") + "/.afl/afl.config.json",
                "/etc/afl/afl.config.json"
        };

        for (String path : searchPaths) {
            if (Files.exists(Path.of(path))) {
                return fromFile(path);
            }
        }

        return fromEnvironment();
    }

    /**
     * Creates a configuration from environment variables.
     */
    public static AgentPollerConfig fromEnvironment() {
        return applyEnvOverrides(defaults());
    }

    private static AgentPollerConfig applyEnvOverrides(AgentPollerConfig cfg) {
        String mongoUrl = Optional.ofNullable(System.getenv("AFL_MONGODB_URL"))
                .filter(s -> !s.isEmpty())
                .orElse(cfg.mongoUrl());
        String database = Optional.ofNullable(System.getenv("AFL_MONGODB_DATABASE"))
                .filter(s -> !s.isEmpty())
                .orElse(cfg.database());
        long pollMs = cfg.pollIntervalMs();
        int maxConc = cfg.maxConcurrent();
        long hbMs = cfg.heartbeatIntervalMs();
        try {
            String v = System.getenv("AFL_POLL_INTERVAL_MS");
            if (v != null && !v.isEmpty()) pollMs = Long.parseLong(v);
        } catch (NumberFormatException ignored) {}
        try {
            String v = System.getenv("AFL_MAX_CONCURRENT");
            if (v != null && !v.isEmpty()) maxConc = Integer.parseInt(v);
        } catch (NumberFormatException ignored) {}
        try {
            String v = System.getenv("AFL_HEARTBEAT_INTERVAL_MS");
            if (v != null && !v.isEmpty()) hbMs = Long.parseLong(v);
        } catch (NumberFormatException ignored) {}

        return new AgentPollerConfig(
                cfg.serviceName(),
                cfg.serverGroup(),
                cfg.serverName(),
                cfg.taskList(),
                pollMs,
                maxConc,
                hbMs,
                mongoUrl,
                database
        );
    }

    /**
     * Creates a new config with a different service name.
     */
    public AgentPollerConfig withServiceName(String serviceName) {
        return new AgentPollerConfig(
                serviceName, serverGroup, serverName, taskList,
                pollIntervalMs, maxConcurrent, heartbeatIntervalMs, mongoUrl, database
        );
    }

    /**
     * Creates a new config with a different server group.
     */
    public AgentPollerConfig withServerGroup(String serverGroup) {
        return new AgentPollerConfig(
                serviceName, serverGroup, serverName, taskList,
                pollIntervalMs, maxConcurrent, heartbeatIntervalMs, mongoUrl, database
        );
    }

    /**
     * Creates a new config with a different task list.
     */
    public AgentPollerConfig withTaskList(String taskList) {
        return new AgentPollerConfig(
                serviceName, serverGroup, serverName, taskList,
                pollIntervalMs, maxConcurrent, heartbeatIntervalMs, mongoUrl, database
        );
    }

    /**
     * Creates a new config with a different max concurrent setting.
     */
    public AgentPollerConfig withMaxConcurrent(int maxConcurrent) {
        return new AgentPollerConfig(
                serviceName, serverGroup, serverName, taskList,
                pollIntervalMs, maxConcurrent, heartbeatIntervalMs, mongoUrl, database
        );
    }
}
